File: 15-3sum/15-3sum.cpp
// Last updated: 2/11/2026, 11:08:42 AM
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(),nums.end());


        for (int i = 0; i < nums.size(); ++i){
            if (i > 0 and nums[i] == nums[i-1]){
                continue;
            }
            int v = nums[i];
            int left = i+1;
            int right = nums.size()-1;

            while (left < right){
                int total = v + nums[left] + nums[right];
                if (total == 0){
                    res.push_back(vector<int>{v,nums[left],nums[right]});

                    left++;
                    right--;

                    while (left < right && nums[left] == nums[left-1]){
                        left++;
                    }

                    while (left < right && nums[right] == nums[right+1]){
                        right--;
                    }
                }

                else if (total > 0){
                    right--;
                }
                else left++;

            }
        }
        return res;
    }

};

File: 15-3sum/15-3sum.py
# Last updated: 2/11/2026, 11:08:44 AM
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        
        
        for idx,num in  enumerate(nums):
            if (idx>0 and num == nums[idx-1]):
                continue
    
            target = -num
            left = idx+1
            right = len(nums)-1
            while (left < right):
                if nums[left] + nums[right] == target:
                    res.append([num, nums[left],nums[right]])
                    left+=1
                    right-=1
                    while left < right and nums[left] == nums[left-1]:
                        left += 1
                    while left < right and nums[right] == nums[right+1]:
                        right -= 1
                elif nums[left] + nums[right] < target:
                    left+=1
                else:
                    right-=1
    

        return res